internal/cli/cmd/cluster/grpcproxy.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	client, err := grpc.DialContext(ctx, "",
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:    time.Second * 30,
+			Timeout: time.Minute,
+		}),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) {
+			return g.connect(ctx)
+		}))
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Fprintf(console.Debug(ctx), "created new grpc connection\n")
+
+	g.backendClient = client
+	return client, nil
+}
+
+func (g *grpcProxy) handler(srv interface{}, serverStream grpc.ServerStream) error {
+	ctx := serverStream.Context()
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		err := status.Errorf(codes.Internal, "reading method failed")
+		fmt.Fprintf(console.Debug(ctx), "reading method failed: %v\n", err)
+		return err
+	}
+
+	fmt.Fprintf(console.Debug(ctx), "handler %s\n", fullMethodName)
+	md, _ := metadata.FromIncomingContext(serverStream.Context())
+	outgoingCtx := metadata.NewOutgoingContext(serverStream.Context(), md.Copy())
+	backendConn, err := g.newBackendClient(outgoingCtx)
+	if err != nil {
+		fmt.Fprintf(console.Debug(ctx), "creating backend connection failed: %v\n", err)
+		return status.Errorf(codes.Internal, "failed to connect to backend: %v", err)
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, clientStreamDescForProxying, backendConn, fullMethodName)
+	if err != nil {
+		fmt.Fprintf(console.Debug(ctx), "failed to create client stream: %v\n", err)
+		return status.Errorf(codes.Internal, "failed create client: %v", err)
+	}
+
+	s2cErrChan := g.proxyServerToClient(serverStream, clientStream)
+	c2sErrChan := g.proxyClientToServer(clientStream, serverStream)
+	// Make sure to close both client and server connections
+	for i := 0; i < 2; i++ {
+		select {
+		case s2cErr := <-s2cErrChan:
+			if s2cErr == io.EOF {
+				clientStream.CloseSend()
+			} else {
+				clientCancel()
+				fmt.Fprintf(console.Debug(ctx), "failed proxying s2c: %v\n", s2cErr)
+				return status.Errorf(codes.Internal, "failed proxying s2c: %v", s2cErr)
+			}
+
+		case c2sErr := <-c2sErrChan:
+			serverStream.SetTrailer(clientStream.Trailer())
+			// c2sErr will contain RPC error from client code. If not io.EOF return the RPC error as server stream error.
+			if c2sErr != io.EOF {
+				fmt.Fprintf(console.Debug(ctx), "failed proxying c2s: %v\n", c2sErr)
+				return c2sErr
+			}
+
+			return nil
+
+		case <-ctx.Done():
+			err := ctx.Err()
+			fmt.Fprintf(console.Debug(ctx), "server stream context is done: %v\n", err)
+			return err
+		}
+	}
+
+	return status.Errorf(codes.Internal, "gRPC proxy should never reach this stage.")
+}
+
+func (g *grpcProxy) proxyClientToServer(src grpc.ClientStream, dst grpc.ServerStream) chan error {
+	ret := make(chan error, 1)
+	go func() {
+		// Not really empty message. All fields are unmarshalled in Empty.unknownFields.
```

Comment:
Ideally we'd close `ret` when the goroutine returns, to avoid leaking a channel.

```
go func() {
  defer close(ret)
}()
```